Add a Windows mode for native commands that allows some commands to use legacy argument passing

JamesWTruher opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This PR targets a _consistent_ default behavior between PowerShell 5 and PowerShell 7 on Windows for some executables and file types. The behavior is now update to use the legacy behavior for the following files:

 - cmd.exe
- cscript.exe
- wscript.exe
- ending with .bat
- ending with .cmd
- ending with .vbs

This list is table driven, and can be easily altered. I have started with this list as I am familiar with real world issues with them.

The PSNativeArgumentPassing preference value has been expanded to include the new value `Windows`.  When the preference variable is set to `Windows` the above table will be in effect, meaning that invocations of those files will automatically use the `Legacy` style argument passing. If the PSNativeArgumentPassing is set to either `Legacy` or `Standard`, then the additional checks will _not_ take place.

The default behavior is now also platform specific. On Windows platforms, the default setting will be `Windows` and non-Windows platforms will be `Standard`.


<!-- Summarize your PR between here and the checklist. -->

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [x] Experimental feature name(s): PSNativeCommandArgumentPassing
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

JamesWTruher committed:
first cut at new arg passing which special cases windows

JamesWTruher committed:
Add validation for the proper behavior based on the platform

JamesWTruher committed:
Updates to address code factor issues.

JamesWTruher committed:
additional code factor fixes

JamesWTruher committed:
try to reduce complexity of GetProcessStartInfo

JamesWTruher committed:
fix spelling error

JamesWTruher committed:
Additional CodeFactor fixes

JamesWTruher committed:
Fix a couple of tests now that the default behavior on Windows has changed

SteveL-MSFT commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -2,6 +2,33 @@
+        $PSNativeCommandArgumentPassing | Should -Be "Standard"
+    }
+    It "Has proper behavior on Windows" -skip:(-not $IsWindows) {
+        "@echo off`nSET V1=1" > "$TESTDRIVE\script 1.cmd"
```
for batch file case, you can have `-testcases` for extension `cmd` and `bat`.  should have tests for .wsf, .vbs, and .js.  You'll probably want to do this before running those tests to set cscript as default (as wscript is the default otherwise), need to run elevated:

```
cscript //h:cscript //nologo //s
```

test.wsf

```xml
<?xml version="1.0" ?>
 <job id="test">
   <script language="VBScript">
     <![CDATA[
for i = 0 to wScript.arguments.count-1
    wscript.echo "Argument " & i & " is: " & wScript.arguments(i)
next
     ]]>
   </script>
 </job>
```

test.vbs

```vbscript
for i = 0 to wScript.arguments.count-1
    wscript.echo "Argument " & i & " is: " & wScript.arguments(i)
next
```

test.js

```javascript
for (i = 0; i < WScript.Arguments.Count(); i++) {
    WScript.echo("Argument " + i + " is: " + WScript.Arguments(i));
}
```

JamesWTruher replied:
@SteveL-MSFT how do you get the `test.js` to run? I can't get it to go without a prepended `cscript`

SteveL-MSFT replied:
`.js` should be associated with wscript by default (hence the cscript command above)

JamesWTruher commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1206,6 +1237,44 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
+                ".cmd",
+                ".vbs"
+                };
+            foreach (string exception in exceptions)
```
Since all I have is the path as a string (not a fileinfo), I'll need to do a string comparison anyway and I can't do "HashSet.Contains (or similar). I sort of feel that creating 2 lists is less efficient than 1.

SteveL-MSFT replied:
This is a full path?  wouldn't that work incorrectly for `mycmd.exe`?

SteveL-MSFT replied:
@JamesWTruher I don't think this comment has been addressed?

JamesWTruher replied:
yes, it's same for cscript.exe and wscript.exe (not sure why, but it would be possible). I will add checks for explicit "\\cmd.exe" and "/cmd.exe"

SteveL-MSFT replied:
I think it would be better to get the FileInfo and have separate hashtable for filename and fileextension.  Alternatively, the exception behavior is only for Windows so we could `#ifdef` that portion and not worry about forward slashes as I would expect the filePath to be normalized by this point?

rjmholt replied:
Rather than try to hand roll this in any particular way, given that we're touching the file system and about to invoke a process, I'd just eat the allocations and do something like this:

```csharp
private static readonly IReadOnlySet<string> s_legacyFileExtensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
{
    ".js",
    ".wsf",
    ".cmd",
    ".bat",
    ".vbs",
}

private static readonly IReadOnlySet<string> s_legacyCommands = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
{
    "cmd",
    "cscript",
    "wscript",
};

private static bool ShouldUseLegacyBehaviour(string commandPath)
{
    ReadOnlySpan<char> extension = Path.GetExtension(commandPath);

    if (s_legacyFileExtensions.Contains(extension))
    {
        return true;
    }

    ReadOnlySpan<char> commandName = Path.GetFileNameWithoutExtension(commandPath);

    if (s_legacyCommands.Contains(commandName))
    {
        return true;
    }

    return false;
}
```

If we can work out how to make the HashSets work with `ReadOnlySpan<char>` we can avoid allocations anyway...

iSazonov replied:
> Alternatively, the exception behavior is only for Windows so we could #ifdef that portion and not worry about forward slashes as I would expect the filePath to be normalized by this point?

NativeCommandProcessor doesn't normalize the path. I am not sure it gets the path normalized.

> I think it would be better to get the FileInfo and have separate hashtable for filename and fileextension.

I'm afraid this array is too small for us to see the benefits of the hashtable/hashset.

rjmholt replied:
```csharp
private static bool ShouldUseLegacyParsing(string commandPath) =>
    !string.IsNullOrEmpty(commandPath)
    && (s_legacyFileExtensions.Contains(Path.GetExtension(commandPath))
        || s_legacyCommands.Contains(Path.GetFileNameWithoutExtension(commandPath)));
```

SteveL-MSFT replied:
@JamesWTruher do you have concerns with changing to Rob's proposal?  I would prefer leveraging the Path methods.

rjmholt replied:
I think what I propose here would still be the right way to go

SteveL-MSFT commented:
I believe this PR addresses part of https://github.com/PowerShell/PowerShell/issues/15143

JamesWTruher committed:
Make sure to change the arguments correctly.

If there is a proxy app which will be executing the app the user specified.
This is really for Windows where the user executes 	est.wsf which will in turn use cscript.exe to do the actual invocation.

JamesWTruher committed:
Add .js and .wsf extensions to the list of specially treated files

JamesWTruher committed:
Add tests for .cmd, .bat, .vbs, .wsf, .js files.

Also will invoke cscript where appropriate.

JamesWTruher committed:
Add additional validation which should help with debugging test failures.

JamesWTruher committed:
Additional debugging output to help determine why it works locally but not in CI

JamesWTruher committed:
Adding explicit call to csript.exe for wsf and vbs files

JamesWTruher commented:
@SteveL-MSFT - I think I've addressed your issues, please let me know

SteveL-MSFT commented:
@JamesWTruher one comment was not addressed

JamesWTruher committed:
Change variable name

to be more representative of what we're working with. Tighten up what will activate the Legacy behavior by including the directory separator for the executables.

iSazonov commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1146,20 +1163,59 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
+        /// Get whether we should treat this executable with special handling and use the legacy passing style.
+        /// </summary>
+        /// <param name="filePath"></param>
+        private bool UseSpecialArgumentPassing(string filePath)
+        {
+            // We need to check if we're using legacy argument passing or it's a special case.
+            bool useLegacy = NativeParameterBinderController.ArgumentPassingStyle == NativeArgumentPassingStyle.Legacy;
+            bool windowsSpecialCase =
+                NativeParameterBinderController.ArgumentPassingStyle == NativeArgumentPassingStyle.Windows &&
+                UseLegacyPassingStyle(filePath);
+            return (useLegacy || windowsSpecialCase);
+        }
```
No need evaluate UseLegacyPassingStyle() if Legacy or not Windows:
```suggestion
        private bool UseSpecialArgumentPassing(string filePath) =>
            NativeParameterBinderController.ArgumentPassingStyle == NativeArgumentPassingStyle.Legacy
            || (NativeParameterBinderController.ArgumentPassingStyle == NativeArgumentPassingStyle.Windows && UseLegacyPassingStyle(filePath));
```

iSazonov commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1206,6 +1263,51 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
+        /// <returns>A boolean indicating what passing style should be used.</returns>
+        private static bool UseLegacyPassingStyle(string filePath)
+        {
+            if (filePath == null || filePath == string.Empty)
+            {
+                return false;
+            }
```
```suggestion
            if (string.IsNullOrEmpty(filePath))
            {
                return false;
            }
```

iSazonov commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1206,6 +1263,51 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
+                return false;
+            }
+
+            string commandPath = filePath.ToLowerInvariant();
```
No need to allocate. We can do `if (filePath.EndsWith(exception, StringComparison.OrdinalIgnoreVase))` in line 1302.

iSazonov commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1206,6 +1263,51 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
+            // The following native commands have non-standard behavior with regard to argument passing.
+            // It's possible (but not likely) that one of the executables could have forward slashes,
+            // so we check for both.
+            string[] exceptions = new string[]
+                {
+                "\\cmd.exe",
+                "/cmd.exe",
```
The array could be static.

rkeithhill commented:
I'm not sure this working completely for .bat files.  preview.6 breaks the `Invoke-BatchFile` command we have implemented in PSCX:
```
PS> invoke-batchfile 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat' amd64
'\"C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat\"' is not recognized as an internal or external command,
operable program or batch file.
```
This is basically Lee Holmes old function that we have implemented in PSCX:
```powershell
function Invoke-BatchFile
{
    param([string]$Path, [string]$Parameters)

    $tempFile = [IO.Path]::GetTempFileName()

    ## Store the output of cmd.exe.  We also ask cmd.exe to output
    ## the environment table after the batch file completes
    cmd.exe /c " `"$Path`" $Parameters && set " > $tempFile
    ...
}
```

vexx32 commented:
@rkeithhill isn't that just because you're passing `$Parameters` as a single string? If you expected multiple args I'd expect that parameter to be `[string[]]` and to pass each parameter separately, myself... though I know passing a space-separated string does sometimes work currently.

JamesWTruher commented:
@rkeithhill wrt the pscx issue that you mentioned. Have you tried running _this_ branch against your scenario? This PR is _supposed_ to automatically put the arg passing into legacy mode for cmd.exe invocations, so it should be fine (but I would really like confirmation on that). I've added validation for that very scenario, but I might have missed something.

JamesWTruher committed:
incorporate feedback from iSazonov

rkeithhill commented:
I built this branch and yes, that command works now while I still have the mode set to `Windows`:

![image](https://user-images.githubusercontent.com/5177512/121964380-b42a7780-cd28-11eb-842c-47e2ce7a0c86.png)

JamesWTruher commented:
@rkeithhill - excellent - the default setting for windows will be "Windows", which is the hybrid model where certain commands will run as Legacy while others will use "Standard". This should hopefully be better and allow the wacky parsing that happens for batch files, but enable more modern apps to use the new style.

SteveL-MSFT reviewed (changes requested):
Please take a look at Rob's suggestion

mklement0 commented:
@SteveL-MSFT 

> I believe this PR addresses part of #15143

I haven't looked at the code, but judging by this PR's description, it doesn't:

* `cmd /c $someCommandLine` (e.g., `cmd /c 'echo "hi there"'`) never needed fixing - _it is the only case that always worked correctly_ - serendipitously, because PowerShell's non-escaping of embedded `"` happens to be canceled out by `cmd.exe`'s "quirky" parsing.
  * Conversely, it mustn't be broken in _any_ mode (it currently is broken in `Standard` mode - see #15239).

* Therefore, none of the truly new accommodations proposed in #15143 are seemingly addressed by this PR.

  * Conversely, selectively reverting to the - hopelessly broken - _legacy_ behavior just means replacing an old mess with a new one (and a more confusing one at that).

---

The only way to avoid a mess is to:

* Stick with two modes: `Legacy` vs. `Standard`.

* In `Standard` mode, implement all accommodations proposed in  #15143 - this should make the vast majority of calls that  (rightfully) rely on PowerShell syntax only just work; for the rest, `--%` or a (non-broken) `cmd /c $someCommandLine` allow full control over the command line passed on Windows.

---

And as for the proposed _default_ values for `$PSNativeCommandArgumentPassing`: 
(I'm aware of the irony of saying this, given that I have often pushed for breaking changes.)

While I'm personally fine with defaulting to the _new_ behavior (assuming it is sensibly implemented), which will undoubtedly make future users happy, existing users with legacy code should brace for a massively breaking change.

SteveL-MSFT commented:
@mklement0 the purpose of this PR is to augment the `NativeCommandArgumentPassing` experimental feature so that `cmd` (and some others) are special cased when using `Windows` mode for argument passing which falls back to legacy behavior.  `Windows` mode will be default on Windows.

The big feedback we got from customers using a PS7 preview in production is that they can't simply move to a brand new arg passing model as they need to support both WinPS5.1 and PS7.  Since 5.1 will still be used quite a bit for a long time, AND we expect users of 5.1 have already worked out workarounds needed that they would expect to continue to work in 7.2, reverting to legacy when using `Windows` and the allow list (cmd, js, etc...) would be the most compatible.

mklement0 commented:
@SteveL-MSFT, i can see only 3 scenarios, and none of them are helped by what this PR proposes - on the contrary:

* (a) Users who are unencumbered with respect to legacy code:
  * They can use `Standard` mode - as modified by #15143 - and live happily ever after.

* (b) Users who use legacy code they cannot update / don't want to spend time updating, but needn't worry about future code they write being backward-compatible with Windows PowerShell:
  * They can selectively use `Legacy` mode - for the legacy code - and live happily ever after.
  * That said, selectively applying a preference variable is hampered by cross-scope-domain code not honoring the caller's preference variables - see #4568 - but that's a separate debate.

* (c) Users who must continue to author code that also runs in Windows PowerShell.

  *  They can use `Legacy` mode consistently - which, as before, is cumbersome and requires obscure workarounds, but that's an unavoidable price to pay.
  * The pain can be mitigated via a helper function that emulates the new `Standard` behavior (as modified by #15143), such as the `ie` function from the [`Native` module](https://github.com/mklement0/Native). (Believe me, this isn't about promoting _my_ module's function; it just so happens to be the least painful solution that I know of; in an ideal world there would never have been a need for such a function.)

Let's be clear that the proposed `Windows` mode will still break a good deal of existing code based on legacy workarounds, namely all calls to executables / interpreters _not_ in the table of exceptions, such as Node.js, for instance.

In `Windows` mode, having to reason with respect to a given call whether the legacy or the new behavior applies only complicates matters - the example I've given before is the `az` CLI, which just so happens to be implemented as `az.cmd`, i.e. as  a _batch file_. But that is an _implementation detail_ that no user should have to pay attention to or should even need to be aware of.

mklement0 commented:
To flesh out my previous concern:

My _guess_ is that many existing workarounds on Windows  do _not_ relate to the executables in the fallback-to-legacy tables, and such workarounds will simply break with `Windows` in effect; e.g.:

```powershell
# *Workaround* that works up to v7.1, but breaks with a `$PSNativeCommandArgumentPassing` value
#  other than 'Legacy'.
curl.exe https://httpbin.org/post -d '{ \"foo\": \"bar\" }' -H 'Content-Type: application/json'

# Ditto.
node -e 'console.log(\"hi\")'
``` 

And, of course, with `Standard` as the default on Unix, _all_ existing workarounds will break - but the real concern is the confusing blending of correct and broken legacy behavior on Windows in the form of the ill-begotten `Windows` mode, depending on how the target executable happens to be implemented.

And even if you _do_ know what behavior applies, that you selectively have to revert to workarounds for _broken_ behavior - even though PowerShell _could_ implement the correct behavior - sounds like a cruel joke.

Again: for someone who needs to continue to author code for both editions, sticking with `Legacy` and the known workarounds _consistently_ is by far the simpler solution.

Someone who needs to selectively run old code can activate `Legacy` selectively (or reverse the logic: activate `Standard` selectively).

That PowerShell lacks lexical scoping for preferences and, conversely, a cross-module mechanism for setting preferences, and therefore makes their selective application challenging is unfortunate, but not specific to the problem at hand.

Ultimately, I think, the only headache-free way out of this mess is via a _distinct call operator or function_ that selects either the legacy or the (consistently) correct behavior: then each call's behavior can easily reasoned about _by itself_, and doesn't depend on possibly hidden state through dynamically scoped preferences.

Of course, the challenges are (a) what operator sigil or function name to come up with, and (b) the separate operator/function serving as a perennial reminder that the original implementation was broken.

Luiz-Monad commented:
This is such a dumpster fire, like I use powershell because I couldn't care less about `cmd`, I just wanted it to die. 
(who writes batch files ? they are all considered broken anyway, but I do understand they need to be supported, just not at the expense of the evolution of Powershell, otherwise I'll just plug my own `NativeCommandProcessor`, wait there's no way to extend custom command processors, that would be a neat feature for solving those nastiness, like create a CmdCommandProcessor already and keep it away from the rest of powershell !)
Please don't break `$PSNativeCommandArgumentPassing` by setting it to other than `Standard` or worse, making it do shenanigans other than passing all the arguments directly to Process-Start args collection. 

```
cinst powershell.core
pwsh
& something 'should just work'
```

All I was doing was just trying to make my CMake powershell build script run multiplatform, with support for VT, and redirecting the error output with performance, but that's another problem, I made my own CmdLet for this on top of the start process, I was looking for something like `Native` to solve this nastiness.

I guess, it would be easier to recompile CMake with a parameter like `EncodedCommand` that passes a single Base64 string.
Perhaps this is the only way...

SteveL-MSFT commented:
@PowerShell/powershell-committee discussed the feedback, we would be concerned about trying to abstract all known workarounds on Windows given the inconsistency of how native commands handle escaping.  The intent of this PR is to address a known impact raised by users of this experimental feature.  In the example of:

```powershell
curl.exe https://httpbin.org/post -d '{ \"foo\": \"bar\" }' -H 'Content-Type: application/json'
node -e 'console.log(\"hi\")'
```

Users who want to maintain known workarounds, they should use `Legacy`.  With `Windows` mode (or `Standard`) they would not need to escape the double quotes:

```powershell
curl.exe https://httpbin.org/post -d '{ "foo": "bar" }' -H 'Content-Type: application/json'
node -e 'console.log("hi")'
```

Based on real world user feedback, we agreed to address some well known and used legacy apps like wscript and cmd for `Windows` mode.  We will evaluate any new feedback as separate issues/discussion from using this experimental feature.

JamesWTruher committed:
Update src/System.Management.Automation/engine/CommandBase.cs

Co-authored-by: Steve Lee <slee@microsoft.com>

SteveL-MSFT commented:
Reassigning to @rjmholt as @daxian-dbw is taking time off right now

rjmholt commented on src/System.Management.Automation/engine/NativeCommandParameterBinder.cs:
```diff
@@ -198,16 +198,16 @@ internal bool UseArgumentList
                         // This will default to the new behavior if it is set to anything other than Legacy
                         var preference = LanguagePrimitives.ConvertTo<NativeArgumentPassingStyle>(
                             Context.GetVariableValue(new VariablePath(SpecialVariables.NativeArgumentPassing), NativeArgumentPassingStyle.Standard));
```
This variable path should be a static readonly so we don't need to allocate a new one every time

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -485,7 +486,16 @@ private void InitNativeProcess()
                             string oldArguments = startInfo.Arguments;
                             string oldFileName = startInfo.FileName;
-                            startInfo.Arguments = "\"" + startInfo.FileName + "\" " + startInfo.Arguments;
+                            bool useSpecialArgumentPassing = UseSpecialArgumentPassing(oldFileName);
+                            // Check to see whether we should be using Legacy mode
```
```suggestion
                            // Check to see whether this executable should be using Legacy mode argument parsing
                            bool useSpecialArgumentPassing = UseSpecialArgumentPassing(oldFileName);
```

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1108,14 +1125,14 @@ private void ProcessOutputRecord(ProcessOutputObject outputValue)
-        /// <param name="soloCommand"></param>
-        /// <returns></returns>
-        private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectError, bool redirectInput, bool soloCommand)
+        /// <param name="redirectOutput">A boolean which determines whether the output is to be redirected.</param>
```
```suggestion
        /// <param name="redirectOutput">If true, redirect output from the process to a stream, otherwise send it to stdout.</param>
```

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1108,14 +1125,14 @@ private void ProcessOutputRecord(ProcessOutputObject outputValue)
-        /// <returns></returns>
-        private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectError, bool redirectInput, bool soloCommand)
+        /// <param name="redirectOutput">A boolean which determines whether the output is to be redirected.</param>
+        /// <param name="redirectError">A boolean which determines whether errors are to be redirected.</param>
```
```suggestion
        /// <param name="redirectError">If true, redirect error output from the process to a stream, otherwise send it to stderr.</param>
```

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1108,14 +1125,14 @@ private void ProcessOutputRecord(ProcessOutputObject outputValue)
-        private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectError, bool redirectInput, bool soloCommand)
+        /// <param name="redirectOutput">A boolean which determines whether the output is to be redirected.</param>
+        /// <param name="redirectError">A boolean which determines whether errors are to be redirected.</param>
+        /// <param name="redirectInput">A boolean which determines whether the input is to be redirected.</param>
```
```suggestion
        /// <param name="redirectInput">If true, take input for the process from a stream, otherwise take it from stdin.</param>
```

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1108,14 +1125,14 @@ private void ProcessOutputRecord(ProcessOutputObject outputValue)
+        /// <param name="redirectOutput">A boolean which determines whether the output is to be redirected.</param>
+        /// <param name="redirectError">A boolean which determines whether errors are to be redirected.</param>
+        /// <param name="redirectInput">A boolean which determines whether the input is to be redirected.</param>
+        /// <param name="soloCommand">A boolean which determines whether this is a solo command.</param>
```
This comment just restates what the variable name already says, but I don't know enough to provide an alternative here. Please change the comment to supplement the information in the variable name.

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1108,14 +1125,14 @@ private void ProcessOutputRecord(ProcessOutputObject outputValue)
+        /// <param name="redirectInput">A boolean which determines whether the input is to be redirected.</param>
+        /// <param name="soloCommand">A boolean which determines whether this is a solo command.</param>
+        /// <returns>A ProcessStartInfo object which is the base of the native invocation.</returns>
+        private ProcessStartInfo GetProcessStartInfoObject(bool redirectOutput, bool redirectError, bool redirectInput, bool soloCommand)
```
I feel the name of this method was previously just as accurate but more succinct; in OOP, the "Object" suffix is usually tautologous, so doesn't add extra information and is unneeded.

```suggestion
        private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectError, bool redirectInput, bool soloCommand)
```

rjmholt replied:
I've just seen that this is a new method to be used in the existing `GetProcessInfo()`, but I'm not sure why it's been factored out as a method at this point.

We should work out what the purpose of each method is (by answering the question "what does this method do?" with a single sentence and without the word "and") and name each one in a way that matches that description.

rjmholt commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1146,20 +1163,53 @@ private ProcessStartInfo GetProcessStartInfo(bool redirectOutput, bool redirectE
+        /// </summary>
+        /// <param name="filePath"></param>
+        private bool UseSpecialArgumentPassing(string filePath) =>
+            NativeParameterBinderController.ArgumentPassingStyle == NativeArgumentPassingStyle.Legacy
+            || (NativeParameterBinderController.ArgumentPassingStyle == NativeArgumentPassingStyle.Windows && UseLegacyPassingStyle(filePath));
```
Might be easier to read like this:

```suggestion
            NativeParameterBinderController.ArgumentPassingStyle switch
            {
                NativeArgumentPassingStyle.Legacy => true,
                NativeArgumentPassingStyle.Windows => UseLegacyPassingStyle(filePath),
                _ => false
            };
```

rjmholt commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -2,7 +2,176 @@
+
+Describe "Behavior is specific for each platform" -tags "CI" {
+    BeforeAll {
+        $skipTests = $false
+        if ($EnabledExperimentalFeatures -notcontains 'PSNativeCommandArgumentPassing') {
+            $skipTests = $true
+        }
```
```suggestion
        $skipTests = $EnabledExperimentalFeatures -notcontains 'PSNativeCommandArgumentPassing'
```

rjmholt commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -2,7 +2,176 @@
+
+Describe "tests for multiple languages and extensions" -tags "CI" {
+    AfterAll {
+        if ($EnabledExperimentalFeatures -notcontains 'PSNativeCommandArgumentPassing') {
+            return
+        }
+        if (! $IsWindows ) {
+            return
+        }
```
```suggestion
        if (-not $IsWindows -or
            $EnabledExperimentalFeatures -notcontains 'PSNativeCommandArgumentPassing') {
            return
        }
```

rjmholt commented on test/powershell/Language/Scripting/NativeExecution/NativeCommandArguments.Tests.ps1:
```diff
@@ -2,7 +2,176 @@
+
+        # determine whether we should skip the tests we just defined
+        # doing it in this order ensures that the test output will show each skipped test
+        $skipTests = $false
+        if ($EnabledExperimentalFeatures -notcontains 'PSNativeCommandArgumentPassing') {
+            $skipTests = $true
+            return
+        }
+
+        if (! $IsWindows ) {
+            $skipTests = $true
+            return
+        }
```
```suggestion
        $skipTests = -not $IsWindows -or $EnabledExperimentalFeatures -notcontains 'PSNativeCommandArgumentPassing'
        if ($skipTests) {
            return
        }
```

rjmholt committed:
Apply suggestions from code review

rjmholt committed:
Fix comments

rjmholt committed:
Rework use-legacy test

rjmholt committed:
Use switch expression

rjmholt committed:
Refactor ProcessStartInfo creation

rjmholt committed:
Fix parameter ordering

rjmholt commented:
Opened https://github.com/JamesWTruher/PowerShell-1/pull/15 to add remaining suggested changes

rjmholt committed:
Use static variable path

JamesWTruher committed:
Merge pull request #15 from rjmholt/NativeArgPassing002

Make suggested changes to NativeArgPassing PR

rjmholt commented:
/azp run
